qualification/general/test_control.py

```diff
@@ -0,0 +1,206 @@
+################################################################################
+# Copyright (c) 2022-2024, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Test that controlling a correlator doesn't cause data to be lost."""
+
+import asyncio
+import time
+from collections.abc import AsyncGenerator, Awaitable
+
+import aiokatcp
+import numpy as np
+import pytest
+from pytest_check import check
+
+from katgpucbf.fgpu.delay import wrap_angle
+
+from ..cbf import CBFRemoteControl
+from ..recv import BaselineCorrelationProductsReceiver, TiedArrayChannelisedVoltageReceiver, XBReceiver
+from ..reporter import Reporter
+
+# These magic values are from MK+ Delay Tracking Requirements draft revision E
+DELAY_INTERVAL = 6.0
+MAX_DELAY = 80e-6
+MAX_DELAY_RATE = 2.56e-9
+MAX_PHASE_RATE = 49.22
+# From M0000-0000-096 rev 1
+BEAM_DELAY_INTERVAL = 3.4
+BEAM_MAX_DELAY = 509e-9
+
+
+async def measure(awaitable: Awaitable) -> float:
+    """Time how long an awaitable takes to await."""
+    loop = asyncio.get_running_loop()
+    start = loop.time()
+    await awaitable
+    return loop.time() - start
+
+
+async def periodically(rng: np.random.Generator, interval: float) -> AsyncGenerator[float, None]:
+    """Yield periodically.
+
+    The first yield happens after a random delay in [0, `interval`) seconds,
+    after which each subsequent yields happens every `interval` seconds.
+
+    This raises an :exc:`AssertionError` if the next scheduled time is missed
+    (already in the past).
+    """
+    loop = asyncio.get_event_loop()
+    await asyncio.sleep(rng.uniform(0, interval))
+    start = loop.time()
+    target = start
+    while True:
+        yield target
+        target += interval
+        now = loop.time()
+        assert now <= target
+        await asyncio.sleep(target - now)
+
+
+async def consume_chunks(receiver: XBReceiver, timestamps: list[int]):
```

Comment:
That's... interesting. I don't know why it's getting ignored. pre-commit ignores the scratch/ directory but qualification tests should be getting checked.

Reply:
It turns out that pydocstyle ignores test*.py by default. I guess that makes sense in that test files are not public API, but I think it would be good to force us to write good documentation for tests too. I'll file a separate ticket for that because there are lots of existing violations and I don't think it makes sense to address them in this PR. However I'll document the functions in this file.

Reply:
I think that's a good idea.